File: latihan/2-inheritance/Minibus.cpp
/*  NIM     : 13521129
    Nama    : Chiquita Ahsanunnisa
    Tanggal : 19 Februari 2023
    Topik   : Latihan Inheritance - Minibus.cpp
*/

#include "Minibus.hpp"

// MINIBUS (DERIVED CLASS)

// CTOR
Minibus::Minibus() : Kendaraan(0, 0, "XXX") {}
Minibus::Minibus(int n, int t, string m) : Kendaraan(n,t,m) {}

// CCTOR tidak butuh karena tidak ada DMA
// Minibus::Minibus(const Minibus& mb) {}

// ASSIGNMENT tidak butuh karena tidak ada DMA
// Minibus& Minibus::operator=(const Minibus& mb) {}

// DTOR tidak butuh karena tidak ada DMA baik di base maupun derived class
// Minibus::~Minibus() {}

int Minibus::diskon(int lamaSewa) {
    return (lamaSewa > 10 ? 0.1 * this->biayaSewa(lamaSewa) : 0);
}
// redefine virtual method
void Minibus::printInfo() const {
    Kendaraan::printInfo();
    cout << "Kategori: Minibus" << endl;
}
int Minibus::biayaSewa(int lamaSewa) const {
    return (lamaSewa <= 5 ? 5000000 : 5000000 + 500000 * (lamaSewa - 5));
}

// VIRTUAL COPY
Kendaraan* Minibus::clone() const {
    return new Minibus(*this);
}

File: latihan/5-java-inheritance/Minibus.java
/*  NIM     : 13521129
    Nama    : Chiquita Ahsanunnisa
    Tanggal : 31 Maret 2023
    Topik   : Latihan Java Inheritance - Minibus.java
*/

public class Minibus extends Kendaraan {
    // CTOR
    public Minibus() { super(); }
    public Minibus(int nomor, int tahun, String merk) {
        super(nomor, tahun, merk);
    }
    public Minibus(Minibus other) { 
        super(other.getNomor(),other.getTahun(),other.getMerk()); 
    }

    // OTHERS
    public int biaya(int lamaSewa) {
        return (lamaSewa <= 5 ? 5000000 : 5000000 + 500000 * (lamaSewa - 5));
    }
    public int diskon(int lamaSewa) {
        return (lamaSewa > 10 ? biaya(lamaSewa) / 10 : 0);
    }
    @Override
    public void printInfo() {
        super.printInfo();
        System.out.println("Kategori       : Minibus");
    }
    @Override
    public int biayaSewa(int lamaSewa) {
        return biaya(lamaSewa) - diskon(lamaSewa);
    }
    @Override
    public Kendaraan copy() { return new Minibus(this); }
}
